mitodl/mitxpro
# Upgrade django-oauth-toolkit to 1.4.0

#### Pre-Flight checklist

**Note:** Description of this PR might be long(apologies) but it would be a good idea to read it to get the better understanding of what actually happened.



#### What are the relevant tickets?
#2122 

#### What's this PR do?
- Upgrades `django-oauth-toolkit` from `1.2.0` to `1.4.0`
- Fixes the migration compatibility issues that came up with upgrade

#### How should this be manually tested?
We can/should test it on both new and existing environment but i'd suggest testing it on existing environment with all the data and migrations applied.
- While building the app through docker or running migration in shell, make sure that you see none of `NodeNotFoundError` or `InconsistentMigrationHistory`. (Related to the package upgrade and their migration squashing)
- Once the you build the app, Assuming you have all the integration of xPRO and edX in place, you can now test logging into the xPRO system and make sure that the user is logged into the edX instance without as well.
- Create a new user, Make sure that the user is created successfully in both xPRO and edX - Now login with this user and make sure logging in works fine.

#### Where should the reviewer start?
Before you start, Try to setup & integrate xPRO and edX before applying this branch.
When starting the testing you will need to see the build migrations with updated package.

#### Any background context you want to provide?
- This change is required in https://github.com/mitodl/mitxpro/pull/2022
- The package in this case (django-oauth-toolkit) had an upgrade where they squashed their old migrations into one, it started to break for us because one of our migration(https://github.com/mitodl/mitxpro/blob/master/courseware/migrations/0001_add_oauth_application.py#L36) dependency had on their's (https://github.com/jazzband/django-oauth-toolkit/blob/1.2.0/oauth2_provider/migrations/0005_auto_20170514_1141.py)
- After the 1.4.0 upgrade they squashed the migrations (https://github.com/jazzband/django-oauth-toolkit/pull/668) which in turn made the migration in our system inconsistent(we started to see `NodeNotFoundError` and `InconsistentMigrationHistory`).

**Applicable Solutions**
**1) Change Dependency in our migration:**
- One of the available options to a fix were, Change the dependency migration to the one which is now available in the package after upgrade. 
- Updating our the dependency to https://github.com/jazzband/django-oauth-toolkit/blob/1.4.0/oauth2_provider/migrations/0003_auto_20201211_1314.py makes our dependencies inconsistent in existing envs since our dependent migration is already applied.
- I've changed it to squashed https://github.com/jazzband/django-oauth-toolkit/blob/1.4.0/oauth2_provider/migrations/0001_initial.py, which seems like best option untill we really need to change the dependency to https://github.com/jazzband/django-oauth-toolkit/blob/1.4.0/oauth2_provider/migrations/0003_auto_20201211_1314.py

**2)Play around with faking the migrations**
This would have needed multiple manual steps to be performed by devs and Devops, i didn't opt for this
